### PR TITLE
[Yul] Extract side-effects into their own struct.

### PR DESCRIPTION
**Pull request by chriseth**

### CODE REVIEW

File: libyul/SideEffects.cpp [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+	GNU General Public License for more details.
+
+	You should have received a copy of the GNU General Public License
+	along with solidity.  If not, see <http://www.gnu.org/licenses/>.
+*/
+/**
+ * Specific AST walkers that collect semantical facts.
+ */
+
+#include <libyul/optimiser/Semantics.h>
```
Comment (ekpyron) [inline libyul/SideEffects.cpp]: I'd consider renaming and restructuring the files a bit - it's getting a bit confusing...
This here should be ``libyul/SideEffectsCollector.cpp`` and maybe a ``libyul/SideEffectsCollector.h`` should be extracted from ``libyul/optimiser/Semantics.h``?

At the moment this is called ``SideEffects.cpp``, but implements nothing declared in ``SideEffects.h`` and one needs to look for the declaration of ``SideEffectsCollector`` until one finds it in ``Semantics.h``...

File: libyul/SideEffects.cpp [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+}
+
+void SideEffectsCollector::operator()(FunctionalInstruction const& _instr)
+{
+	ASTWalker::operator()(_instr);
+
+	if (!eth::SemanticInformation::movable(_instr.instruction))
+		m_sideEffects.movable = false;
+	if (!eth::SemanticInformation::sideEffectFree(_instr.instruction))
+		m_sideEffectFree = false;
```
Comment (ekpyron) [inline libyul/SideEffects.cpp]: Now I'm confused - am I looking at the right header for this? Aren't these members removed there and it should access indirectly via ``m_sideEffects`` now as above for ``movable``? If so why does this PR even build? I'm stopping my review here for now, until this is clarified..

File: libyul/SideEffects.cpp (hunk 1 above)
Comment (ekpyron) [reply libyul/SideEffects.cpp]: The reason is that this file is not in the ``CMakeLists.txt``, so it won't be built (it's not used in this PR yet it seems)...

**Review by ekpyron — CHANGES_REQUESTED** [review]

**Comment by chriseth** [comment]
@ekpyron I'm sorry, that file (SideEffects.cpp) was not meant to be part of this PR.

**Comment by chriseth** [comment]
Removed SideEffects.cpp

File: libyul/SideEffects.h [hunk 3]
```diff
@@ -0,0 +1,78 @@
...
+ * Side effects of code.
+ *
+ * The default-constructed value applies to the "empty code".
+ */
+struct SideEffects
+{
+	/// If true, expressions in this code can be freely moved and copied without altering the
+	/// semantics.
+	/// At statement level, it means that functions containing this code can be
+	/// called multiple times, their calls ran be rearranged and calls can also be
```
Comment (leonardoalt) [inline libyul/SideEffects.h]:
```suggestion
	/// called multiple times, their calls can be rearranged and calls can also be
```
?

File: libyul/SideEffects.h [hunk 4]
```diff
@@ -0,0 +1,78 @@
...
+	/// deleted without changing the semantics.
+	/// This means it cannot depend on storage or memory, cannot have any side-effects,
+	/// but it can depend on state that is constant across an EVM-call.
+	bool movable = true;
```
Comment (leonardoalt) [inline libyul/SideEffects.h:39]: Why do they now start with the best case side effects?

**Comment by leonardoalt** [comment]
lgtm apart from typo and question

File: libyul/SideEffects.h (hunk 4 above)
Comment (chriseth) [reply libyul/SideEffects.h:39]: The default-constructed side-effects correspond to "empty code".

**Review by ekpyron — APPROVED** [review]

**Pull request merged**